### PR TITLE
Pass all http headers in request metadata

### PR DESCRIPTION
**Pull request by steve-nester-uk**
PLAT-4629

## Goal

The notifier currently only append headers that start with HTTP_ prefix.  Instead all headers should be sent.

## Changeset

Update the BasicResolver to use getallheaders if it exists, this will set all HTTP headers

## Tests

Update to existing RequestTest.php

## Review

For the submitter, initial self-review:

- [ x] Commented on code changes inline explain the reasoning behind the approach
- [ x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ x] Initial review of the intended approach, not yet feature complete
  - [ x] Structural review of the classes, functions, and properties modified
  - [ x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: src/Request/BasicResolver.php [hunk 1]
```diff
@@ -58,10 +58,17 @@ protected static function getRequestHeaders(array $server)
 
         $headers = [];
 
-        foreach ($server as $name => $value) {
-            if (substr($name, 0, 5) == 'HTTP_') {
-                $headers[str_replace(' ', '-', ucwords(strtolower(str_replace('_', ' ', substr($name, 5)))))] = $value;
+        if (!function_exists('getallheaders')) {
```
Comment (imjoehaines) [inline src/Request/BasicResolver.php:61]: Small nitpick — if there's an else then I usually like the "positive" case first, i.e.

```php
if (function_exists('getallheaders')) {
    $headers = getallheaders();
} else {
    // ...
}
```

We could even avoid the else entirely:

```php
if (function_exists('getallheaders')) {
    $headers = getallheaders();

    return $headers;
}

$headers = [];

foreach ($server as $name => $value) {
    // ...
}
```

File: src/Request/BasicResolver.php [hunk 2]
```diff
@@ -58,10 +58,17 @@ protected static function getRequestHeaders(array $server)
...
+            foreach ($server as $name => $value) {
+                if (substr($name, 0, 5) == 'HTTP_') {
+                    $headers[str_replace(' ', '-', ucwords(strtolower(str_replace('_', ' ', substr($name, 5)))))] = $value;
+                }
+                else if ($name === "CONTENT_TYPE") {
```
[the remark this answers is not in the document]
Comment (steve-nester-uk) [reply src/Request/BasicResolver.php:66]: getallheaders does seem to be called on some of the test instances because before I put in 'fix failures on PHP 5.5.9' fix, most tests were successful except 5.5.9 - those were not setting Content-Type, so must have been using getallheaders.

[the remark this answers is not in the document]
Comment (steve-nester-uk) [reply src/Request/BasicResolver.php:66]: Fixed in 033bd8932af57e286b612d02b715b7efa598f323

**Comment by GrahamCampbell** [comment]
We have to be careful that we're not sending sensitive data now, by default. We should redact the `Authentication` header, and perhaps other custom headers, maybe those with keys that contain "auth" or "key".

**Comment by imjoehaines** [comment]
> We have to be careful that we're not sending sensitive data now, by default. We should redact the `Authorization` header, and perhaps other custom headers, maybe those with keys that contain "auth" or "key".

What extra headers are included by `getallheaders` that aren't in `$_SERVER` as `HTTP_x`? `Authorization` is `$_SERVER['HTTP_AUTHORIZATION']`, for example, so it's already being sent by default

**Comment by GrahamCampbell** [comment]
Oh, maybe that behaviour should be revisited. ;)

**Comment by imjoehaines** [comment]
> Oh, maybe that behaviour should be revisited. ;)

Yeah, we filter those headers on other platforms so should bring PHP in line for sure 🙂
